crates/red_knot_python_semantic/docs/kind_of_types.md

```diff
@@ -0,0 +1,201 @@
+that the type is a singleton type.
+
+In order for a type to be considered a single-value type,
+there must exist a reflexive, symmetric, and transitive equivalence relation
+between all inhabitants of the type. It follows that for a single-value type in Python,
+all inhabitants must be instances of the same runtime class,
+and the class in question must have `__eq__` and `__ne__` methods
+such that the equality relation between instances of the class
+is reflexive, symmetric, and transitive.
+
+Single-value types in unions can be safely narrowed using inequality (but not using equality):
+
+```py
+from typing import Literal
+
+def f(x: str | Literal[1]):
+    if x != 1:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+    else:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+
+    if x == 1:
+        ...  # type of `x` is still `str | Literal[1]`
+             # (`x` could be an instance of a `str` subclass that overrides `__eq__`
+             # to compare equal with `1`)
+    else:
+        ...  # `x` can be narrowed to `str` if not equal to `1`
+```
+
+## Closed types
+
+A closed type is a type for which it is known that the type has no subtypes
+except for the type itself and `Never`; it is "closed for extension".
+For a given closed type `X`, the only subtypes of `X` are `X` and `Never`;
+the only *proper* subtype of `X` is `Never`.
+
+All singleton types and single-value types are closed types.
+However, not all closed types are singleton types or single-value types.
+
+Closed types are often associated with the `@final` decorator.
+Decorating a class with `@final` indicates to the type checker
+that the class should never be subclassed; the type checker will emit a diagnostic
+if it sees it being subclassed, and is free to consider the class as closed for extension.
+As such, the following `@final` class `Spam` creates a corresponding closed type `Spam`:
+
+```py
+from typing import final
+
+@final
+class Spam: ...
+```
+
+However, runtime subclassability does not correspond exactly to whether a type is closed or not.
+For example, the `bool` *class* is (correctly) decorated with `@final` in typeshed's stubs,
+indicating that it cannot be subclassed at runtime. However, the `bool` *type*
+has two proper subtypes other than `Never`: `Literal[True]` and `Literal[False]`.
+As such, although all its subtypes are closed, `bool` itself cannot be considered a closed type.
+Similarly, attempting to subclass the `Eggs` class in the following example
+will lead to an exception at runtime, and type checkers should treat all
+enum classes as implicitly `@final`. Nonetheless, the `Eggs` *type* cannot be considered closed,
+as it has three proper subtypes other than `Never`:
+`Literal[Eggs.A]`, `Literal[Eggs.B]` and `Literal[Eggs.C]`.
```

Comment:
I don't think this paragraph is right. I don't think "has no proper subtypes other than itself and Never" is a workable or useful definition of what it means to be a closed type, and I think both `bool` and `Eggs` should be considered closed types.

Reply:
What _is_ the term you would use for a type where it is known that no proper subtypes exist other than `Never`?

Reply:
I think that would have to be a singleton type. Any type with more than one inhabitant in principle has proper subtypes that are not `Never`.

Reply:
None that we have (or ever would) define or attempt to name, but sure, in principle you can have the singleton type consisting only of the integer object with value `50000000000000000` that exists at a certain memory address, which is a proper subtype of `Literal[50000000000000000]` that is not `Never`.

Reply:
grrr, fine!

Reply:
In some ways I find this a useful clarification: it's good to remember that the set of Python types recognised by red-knot is a subset of the set of possible types that exist in Python. From this perspective, there really exists an infinite number of possible singleton types in Python.

In other ways, I find this not so useful, because I still feel like we're missing a term to describe types which _in red-knot's model_ can have no possible non-`Never` proper subtypes _that would be understood as such by red-knot_. You've successfully convinced me that this is a concept that cannot be expressed in terms of an abstract framework divorced from the concept of Python types and our internal model of Python types. But I still think it's a useful concept nonetheless and something we should have a term for.

Reply:
I think I'm not understanding _why_ this is a useful or important concept.

Reply:
It is what I was _trying_ to describe... but perhaps you're right, maybe it just isn't actually a useful thing to describe. Not sure...

Anyway, I'm reworking the section titled closed types to more accurately describe the actual type-theory concept described by the term "closed types" :-) will probably finish off that rewrite tomorrow morning and push those changes.

Reply:
Ok, looking forward to it!

Reply:
> > `Literal[50000000000000000]` is not a singleton type (multiple inhabitants exist at runtime), but has no non-`Never` proper subtypes
>
> None that we have (or ever would) define or attempt to name, but sure, in principle you can have the singleton type consisting only of the integer object with value `50000000000000000` that exists at a certain memory address, which is a proper subtype of `Literal[50000000000000000]` that is not `Never`.

This is purely a discussion about nomenclature, so I'm not sure how useful it is to go deeper, but I'm not sure I completely agree with this *"every possible set of runtime values corresponds to *some* type, even if we can't name/construct it"* viewpoint. I think the following definition is more useful: A type is a syntactic form that we assign to terms of the programming language. In this definition, it's still true that every type represents a set of possible runtime values. But it's not true that every possible set of runtime values corresponds to a type. *Which* particular sets of runtime values can be represented as a type depends on the type system and its expressiveness. In C's type system, you can not assign a type to the set `{true}`. In Python, you can; In Python, you can not assign a type to objects that are used exactly once. In languages with linear types, you can.

In Python's type system, there is no way to construct a type that represents all objects at memory location 0xf7…, and I don't think it's useful to think of types in terms of sets of runtimes values that could *potentially* be represented by a more expressive type system. I would not claim that C has intersection types, just because I can think of a set of runtime values that would be represented by `int8_t & uint8_t = {0, 1, …, 127}`. And I would not claim that Python has linear types, just because I can think of the set of runtime objects that are used exactly once.

Reply:
> This is purely a discussion about nomenclature

Hey, the whole PR is about nomenclature 😆

One complication here is that red-knot's model of Python's types _is_ more expressive than the model outlined by the typing spec. (In fact, you could argue that all Python type checkers necessarily have a more expressive model than the one outlined in the typing spec -- but I think red-knot goes further than most here.) One obvious example is our concept of function-literal and class-literal types -- you _could_ argue that the concept of class-literal types in particular is strongly implied by much of the rest of the typing spec, but nowhere is it explicitly specified. Similarly, we treat `IntersectionType`s as first-class types and infer them in many places -- but an `Intersection` type is still unspecified in the typing spec. (Mypy, for example, [does narrowing in many places that clearly _implies_ the existence of an intersection type](https://github.com/python/mypy/blob/60d1b3776229d3e3c332413ef49ad89be469a5e4/mypy/checker.py#L5416-L5440), but it has no first-class notion of intersection types in its model.)

So I could put forward at least three possible definitions of "type" here:
- A maximally broad definition: "Any theoretical set of runtime objects constitutes a 'type'"
- A maximally narrow definition: "Only a type explicitly specified by the typing spec can be described as a type"
- A maximally useful definition: "Any set of runtime objects constitutes a type if it would plausibly be useful to represent it as a type in red-knot in order to accurately and precisely model Python's runtime semantics"

I think the maximally useful definition is probably the most, well... useful... but it's also the least precise! It's hard to know which sets of runtime objects we might plausibly find useful to represent as types _at some point_.

Reply:
Once a type system has full set-theoretic types (arbitrary unions, intersections, and negations), it becomes quite hard to draw a clear line between which possible sets of objects are a type and which are not; many quite complex types can be constructed "out of thin air". Because of this, and because of what Alex says (which sets of objects we choose to represent as a type internally is not set in stone by either the language or the typing spec, but is essentially up to our own judgment of what is useful for representing Python semantics, and may change over time), I think the clearest approach in our terminology is to avoid defining types based on an assumption that no subtype of the type could ever exist (despite the type having multiple distinguishable inhabitants.)

(But perhaps more important to the original intent of this thread is that I also don't find that definition _otherwise_ useful or correct: I do think that `bool` and `Eggs` are closed types.)

Reply:
> I do think that `bool` and `Eggs` are closed types

Yeah, you're right that I was trying to use the word "closed" to describe a different concept to what it means everywhere else in type theory; that was obviously wrong and silly. No disagreements on _that_ point.